src/rust/src/rustzcash.rs

```diff
@@ -514,24 +457,25 @@ pub extern "system" fn librustzcash_sapling_ka_agree(
     sk: *const [c_uchar; 32],
     result: *mut [c_uchar; 32],
 ) -> bool {
-    // Deserialize p
-    let p = match edwards::Point::<Bls12, Unknown>::read(&(unsafe { &*p })[..], &JUBJUB) {
-        Ok(p) => p,
-        Err(_) => return false,
+    let esk = {
+        let esk = EphemeralSecretKey::from_bytes(&unsafe { *sk });
+        if esk.is_some().into() {
+            esk.unwrap()
+        } else {
+            return false;
+        }
     };
-
-    // Deserialize sk
-    let sk = match Fs::from_repr(read_fs(&(unsafe { &*sk })[..])) {
-        Ok(p) => p,
-        Err(_) => return false,
+    let p = {
+        let p = DiversifiedTransmissionKey::from_bytes(&unsafe { *p });
+        if p.is_some().into() {
+            p.unwrap()
+        } else {
+            return false;
+        }
     };
 
-    // Compute key agreement
-    let ka = sapling_ka_agree(&sk, &p);
-
-    // Produce result
-    let result = unsafe { &mut *result };
-    ka.write(&mut result[..]).expect("length is not 32 bytes");
+    let shared_secret = SaplingDomain::<TestNetwork>::ka_agree_enc(&esk, &p);
+    *&mut unsafe { *result } = shared_secret.0.to_bytes();
```

Comment:
trivial bug (in this way result is not updated)
a possible fix is:
`unsafe{*result = shared_secret.0.to_bytes()};`

Reply:
It's weird that it wouldn't work. I've applied your suggestion.